Foundation/FileHandle.swift

```diff
@@ -31,102 +31,95 @@ open class FileHandle : NSObject, NSSecureCoding {
     }
 
     open var availableData: Data {
-        return _readDataOfLength(Int.max, untilEOF: false)
+        do {
+            let readResult = try _readDataOfLength(Int.max, untilEOF: false)
+            return readResult.toData()
+        } catch {
+            fatalError("\(error)")
+        }
     }
     
     open func readDataToEndOfFile() -> Data {
         return readData(ofLength: Int.max)
     }
 
     open func readData(ofLength length: Int) -> Data {
-        return _readDataOfLength(length, untilEOF: true)
+        do {
+            let readResult = try _readDataOfLength(length, untilEOF: true)
+            return readResult.toData()
+        } catch {
+            fatalError("\(error)")
+        }
     }
 
-    internal func _readDataOfLength(_ length: Int, untilEOF: Bool) -> Data {
+    internal func _readDataOfLength(_ length: Int, untilEOF: Bool, options: NSData.ReadingOptions = []) throws -> NSData.NSDataReadResult {
         precondition(_fd >= 0, "Bad file descriptor")
+        if length == 0 && !untilEOF {
+            // Nothing requested, return empty response
+            return NSData.NSDataReadResult(bytes: nil, length: 0, deallocator: nil)
+        }
+
         var statbuf = stat()
-        var dynamicBuffer: UnsafeMutableRawPointer? = nil
-        var total = 0
         if fstat(_fd, &statbuf) < 0 {
-            fatalError("Unable to read file")
+            throw NSError(domain: NSPOSIXErrorDomain, code: Int(errno), userInfo: nil)
         }
-        if statbuf.st_mode & S_IFMT != S_IFREG {
-            /* We get here on sockets, character special files, FIFOs ... */
-            var currentAllocationSize: size_t = 1024 * 8
-            dynamicBuffer = malloc(currentAllocationSize)
-            var remaining = length
-            while remaining > 0 {
-                let amountToRead = min(1024 * 8, remaining)
-                // Make sure there is always at least amountToRead bytes available in the buffer.
-                if (currentAllocationSize - total) < amountToRead {
-                    currentAllocationSize *= 2
-                    dynamicBuffer = _CFReallocf(dynamicBuffer!, currentAllocationSize)
-                    if dynamicBuffer == nil {
-                        fatalError("unable to allocate backing buffer")
+
+        let readBlockSize: Int
+        if statbuf.st_mode & S_IFMT == S_IFREG {
+            // TODO: Should files over a certain size always be mmap()'d?
```

Comment:
Immediate answer: no, unless if options contains `.alwaysMapped` or `.mappedIfSafe`.

(I don't see `.mappedIfSafe` handled here at all? Are you just assuming mapping is always unsafe? This is allowed by the contract but we should have a follow-up.)

Reply:
`.mappedIfSafe` wasn't handled in the previous code so I haven't added it but  I can do a separate PR for it.